File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-3.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {    
    public bool IsSameTree(TreeNode p, TreeNode q)
    {
        if (p?.val != q?.val)
            return false;

        if (p != null && q != null)
        {
            if (!IsSameTree(p.right, q.right))
                return false;

            if (!IsSameTree(p.left, q.left))
                return false;
        }

        return true;
    }

    public bool IsSubtree(TreeNode root, TreeNode subRoot)
    {
        if (root?.val == subRoot?.val)
        {
            if (IsSameTree(root.right, subRoot.right)
                && IsSameTree(root.left, subRoot.left)
                )
                return true;
        }

        if (root != null)
            return IsSubtree(root.right, subRoot) 
                    || IsSubtree(root.left, subRoot);

        return false;
    }
}

File: Data Structures & Algorithms/valid-binary-search-tree/submission-3.py
class Solution:
    def dfs(self, node, min, max):

        result = True
        while node and result:
            if(node.val <= min or node.val >= max):
                return False
            
            if node.right:
                result = self.dfs(node.right, node.val, max)

            if result and node.left:
                result = self.dfs(node.left,  min, node.val)

            break

        return result


    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        if(root == None):
            return False

        negative_infinity = float('-inf')
        positive_infinity = float('inf')

        return self.dfs(root, negative_infinity, positive_infinity)   